[clement559] iP

clement559 opened the pull request.

clement559 committed:
Added welcome message for Duke as per Week 2 project timeline.

clement559 committed:
Added Level 1 increment to greet, echo and exit.

clement559 committed:
Added Level 2 increment to Add, List. Program now remembers the input items and can output the items in the list using the command "list".

clement559 committed:
Added Level 3 increment, Mark as Done.
Changed list check from direct string check to contains check to avoid adding as task if there are accidental spaces.

clement559 committed:
Added Level 4 increment, ToDos, Events, Deadlines.
Added files Deadline, Event and Todo to subclasses to Task.
Added new function getTypeIcon to Task.

clement559 committed:
Added A-TextUITesting.
Auto tests 1 todo, 1 deadline and 1 event task and the list function.

erjunze commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,93 @@
-        String logo = " ____        _        \n"
+        String lineBreak = "---------------------------------------------";
+        String line = "";
+        Task[] list = new Task[100];
```
Perhaps use a named constant instead of a magic number

woowenjun99 replied:
Code Standards: Perhaps this should be plural to represent a collection of objects?

```suggestion
        int MAX_TASK = 100;
        Task[] lists = new Task[MAX_TASK];
```

Chick3nBoy replied:
Could maybe use Tasks instead of Task for array naming.

erjunze commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,93 @@
+        int index = 0;
+        Scanner in = new Scanner(System.in);
+        /*String logo = " ____        _        \n"
                 + "|  _ \\ _   _| | _____ \n"
                 + "| | | | | | | |/ / _ \\\n"
                 + "| |_| | |_| |   <  __/\n"
                 + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        System.out.println("Hello from\n" + logo);*/
```
Since dead code can be retrieved using revision control, consider deleting it instead of just commenting it out

erjunze commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,93 @@
     public static void main(String[] args) {
-        String logo = " ____        _        \n"
+        String lineBreak = "---------------------------------------------";
+        String line = "";
```
Since you only use this variable inside the while block below, could it be moved inside to minimise its scope?

erjunze commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,93 @@
+            String[] words = line.split(" ");
+            System.out.println(lineBreak);
+            if (line.contains("list")) {
+                System.out.println("Here are the tasks in your list:");
+                int itemNumber = 1;
+                for (Task item : list) {
+                    if ((itemNumber - 1) == index) {
+                        break;
+                    }
+                    System.out.println(itemNumber + ".[" + list[itemNumber - 1].getTypeIcon() + "]"
+                            + "["  + list[itemNumber - 1].getStatusIcon() + "] " + list[itemNumber - 1].description);
+                    itemNumber++;
+                }
+                System.out.println(lineBreak);
```
Could this code be abstracted out so the main method is not so long?

erjunze commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,93 @@
+                break;
+            }
+            else {
+                Task item;
```
Since this variable is only used in the if else blocks below, could it be moved inside to minimise its scope?

Chick3nBoy replied:
Perhaps declare Task item above so that it can be used in the if statement below as well.

erjunze commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,93 @@
+            }
+            else {
+                Task item;
+                if (line.startsWith("todo")) {
+                    String[] inputLine = line.split(" ", 2);
+                    item = new Todo(inputLine[1]);
+                    list[index] = item;
+                    index++;
+                    System.out.println("Got it. I've added this task: " + "[" + item.getTypeIcon() + "]"
+                            + "["  + item.getStatusIcon() + "] " + item.description);
+                }
+                else if (line.startsWith("deadline")) {
+                    String[] inputLine = line.split(" ", 2);
+                    inputLine = inputLine[1].split("/by ");
+                    String description = inputLine[0];
+                    String deadline = inputLine[1];
+                    item = new Deadline(description, deadline);
+                    list[index] = item;
+                    index++;
+                    System.out.println("Got it. I've added this task: " + "[" + item.getTypeIcon() + "]"
+                            + "["  + item.getStatusIcon() + "] " + item.description);
+                }
+                else if (line.startsWith("event")) {
+                    String[] inputLine = line.split(" ", 2);
+                    inputLine = inputLine[1].split("/from ");
+                    String description = inputLine[0];
+                    inputLine = inputLine[1].split("/to ");
+                    String start = inputLine[0];
+                    String end = inputLine[1];
+                    item = new Event(description, start, end);
+                    list[index] = item;
+                    index++;
+                    System.out.println("Got it. I've added this task: " + "[" + item.getTypeIcon() + "]"
+                            + "["  + item.getStatusIcon() + "] " + item.description);
```
Once again, consider abstracting this code out into multiple methods so each method is not so long

erjunze commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,32 @@
+        this.isDone = false;
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
+    public String getTypeIcon() {
+        return "";
+    }
```
Perhaps you can combine the square brackets into these methods so you don't have to repeat them for each of the different tasks, for example:
```
public String getStatusIcon() {
    return (isDone ? "[X]" : "[ ]");
}
```

erjunze commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,32 @@
+        return "";
+    }
+
+    public void markAsDone() {
+        this.isDone = true;
+        System.out.println("Nice! I've marked this task as done:");
+        System.out.println(" [" + this.getStatusIcon() + "] " + this.description);
+        System.out.println(lineBreak);
+    }
+
+    public void markAsUndone() {
+        this.isDone = false;
+        System.out.println("Ok, I've marked this task as not done yet:");
+        System.out.println(" [" + this.getStatusIcon() + "] " + this.description);
+        System.out.println(lineBreak);
+    }
```
Since these two methods are mostly the same, could you combine them to reduce code duplication?

erjunze reviewed (commented):
Overall, there are many improvements that can be made. Focus on splitting up into multiple smaller methods so they are not so long.

woowenjun99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,93 @@
 public class Duke {
     public static void main(String[] args) {
-        String logo = " ____        _        \n"
+        String lineBreak = "---------------------------------------------";
```
Coding Standard: Since this is a constant, you have to capitalise the full name.

```suggestion
        String LINEBREAK = "---------------------------------------------";
```

woowenjun99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,93 @@
+            else {
+                Task item;
+                if (line.startsWith("todo")) {
+                    String[] inputLine = line.split(" ", 2);
```
Coding Standards: Perhaps this should be plural to represent a collection of objects? Furthermore, I think this variable could be better named as the method seemed to breaking up the line into an array of string.

```suggestion
                    String[] inputLines = line.split(" ", 2);
```

Chick3nBoy replied:
Perhaps the naming should be plural i.e. inputLines

woowenjun99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,93 @@
+import java.util.Scanner;
+import java.util.Arrays;
```
Good job on not importing everything!

woowenjun99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,93 @@
+        while (true) {
+            line = in.nextLine();
+            String[] words = line.split(" ");
+            System.out.println(lineBreak);
```
Good job on using plural for the array here!

woowenjun99 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,13 @@
+public class Event extends Task {
```
Good job on naming the class in capital letter

woowenjun99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,93 @@
+                    System.out.println("Got it. I've added this task: " + "[" + item.getTypeIcon() + "]"
+                            + "["  + item.getStatusIcon() + "] " + item.description);
+                }
+                else if (line.startsWith("event")) {
```
Good job on using the K&R style for the if-else block

woowenjun99 reviewed (commented):
Overall good job on following the coding standards. However, I feel that the naming could be better improved for some of the variables, such as using `LINE_BREAK` instead of `lineBreak` for constants. Furthermore, for a collection of objects, the variables should be named in plural and better named, especially the `String inputLine = line.split(" ");`. Otherwise, most of the code is well written and abide to the standards.

Chick3nBoy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,93 @@
+            line = in.nextLine();
+            String[] words = line.split(" ");
+            System.out.println(lineBreak);
+            if (line.contains("list")) {
```
Perhaps a switch statement would be better instead of multiple if else since there are multiple conditions.

Chick3nBoy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,93 @@
+                            + "["  + item.getStatusIcon() + "] " + item.description);
+                }
+                else if (line.startsWith("event")) {
+                    String[] inputLine = line.split(" ", 2);
```
The code associated for event could perhaps be made easier to read by refactoring into a doEvent(); method

Chick3nBoy reviewed (commented):
Overall minimal mistakes, good job!

clement559 committed:
Added A-CodeQuality increment, refactored code and implemented most changes as commented by fellow peers during PR.
Duke main function should still be furthered streamlined for fewer lines of code. Will continue to work on that.

clement559 committed:
Added Level 5 and A-Exceptions increment.
Added exception handler for incorrect number of parameters given for the different type of tasks.
Added exception handler for exiting program when it is still waiting for input.
Added exception for no task type specified.

clement559 committed:
Merge branch 'branch-Level-5'

hughjazzman commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,133 @@
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+    public static final int MAX_NUM_OF_TASKS = 100;
+    public static final String LINE_BREAK = "---------------------------------------------";
```
Perhaps you can use the `String.repeat` method here?

hughjazzman commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,133 @@
+
+    public static void registerDeadline (Task[] lists, String line, int index) throws IndexOutOfBoundsException {
+        try {
+            String[] inputLines = line.split(" ", 2);
+            inputLines = inputLines[1].split("/by ");
```
Might there be some magic literals and numbers here?

hughjazzman commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,133 @@
+import java.util.NoSuchElementException;
+import java.util.Scanner;
+
 public class Duke {
```
You could consider moving `Ui` and `Command` code to separate classes to better apply OOP principles.

hughjazzman commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,133 @@
+            String line = getInput();
+            String[] words = line.split(" ");
+            System.out.println(LINE_BREAK);
+            if (line.contains("list")) {
```
Consider abstracting similar code logic here.

hughjazzman reviewed (commented):
Great effort! You can consider using more classes to apply OOP principles and techniques.

clement559 committed:
**Duke.java**
Imported ArrayList package. Changed array of tasks to ArrayList of tasks. Removed use of constant MAX_NUM_OF_TASKS.
Renamed printTask to printAddTask for less confusion between deletion and addition.
Added exception handler for main, IndexOutOfBounds exception when mark/unmark/delete are not given index numbers of elements currently in the ArrayList. Added printError for the exception handler.
Removed use of int index in main code. Checks for size of ArrayList in the individual functions when needed instead.
Added Level-6 increment to support deletion of tasks.

**Task.java**
Removed additional square brackets in intermediary printing as it has been added to getStatusIcon function instead.

clement559 committed:
Added Level 7 increment to allow saving of current list and reading of existing list in Duke.txt.
Program will save list to /data/Duke.txt. Directory will be made if non-existent.
Changed Array[] lists to ArrayList<Task> lists for better management.

clement559 committed:
Merge branch 'branch-Level-6'

clement559 committed:
Merge branch 'branch-Level-7'

clement559 committed:
Packaged Task and Task types into duke.task
Packaged Main and Storage into duke.util
Pacakged DukeException into duke.exception

Data file now gets saved in src/main/java/data

clement559 committed:
Moved a lot of code to separate files for MORE OOP.
Duke rebrought out of package.

clement559 committed:
Added Level 9 increment to be able to search for items in the existing list.
Amended command check for list to de-conflict with search terms including "list".

clement559 committed:
Added JavaDoc to most functions and classes.
Moved getInput from ui package to parser package.
Removed exception handler when registering to-do task.

clement559 committed:
Merge pull request #1 from clement559/branch-Level-9

Week 7 Stuff

clement559 committed:
Merge branch 'master' into A-JavaDoc

# Conflicts:
#	src/main/java/duke/util/TaskList.java
#	src/main/java/duke/util/Ui.java

clement559 committed:
Merge pull request #2 from clement559/branch-A-JavaDoc

Week 7 Stuffs Pt 2

clement559 committed:
Update README.md

clement559 committed:
Adding more documentation for new functions for search.

clement559 committed:
Creation of new Command package for stretch goal.
Created separate extensions of main Command file for respective commands.
Amended print output for deletion of tasks.

clement559 committed:
Update README.md

Reverting changes to main README file

clement559 committed:
Update README.md

clement559 committed:
Updated printing format to standardise across all commands.

clement559 committed:
Merge branch 'master' of https://github.com/clement559/ip

clement559 committed:
Removed UI object from MarkTaskCommand.
Created new function MarkAsDoneSilent in Task so that when doing initial reading of tasks from Dukefile, no printing occurs when marking completed tasks.

clement559 committed:
Added Level-8 increment.
Edited error messages produced by Deadline, Event and Todo.
Added invalid checker to ensure null tasks do not get created.

clement559 committed:
Merge pull request #3 from clement559/branch-Level-8

Added Level-8 increment.

clement559 committed:
Merge branch 'master' of https://github.com/clement559/ip

clement559 committed:
Updated README.md to reflect Level 8 increment changes.

clement559 committed:
Removed redundant toExit function and isExit variable since deprecated.
Checks are now done in command class.
